File: packages/provide-uterm-ts/src/transports/telnet-transport.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * A full RFC 854 telnet client.
 *
 * Port of the Python module `provide.uterm.transports.telnet_transport`.
 *
 * The framing lives in `./telnet.ts`; this is the conversation. A client that
 * opens without offering anything gets a line-mode, non-binary session and a
 * BBS then draws for the wrong terminal — so the offer goes out first, and
 * each answer that agrees to something is followed immediately by the value
 * the far end needs (the window size, the terminal type), because agreeing
 * and then staying silent leaves the server guessing.
 */

import { type ConnectionTransport, TransportConnectionError } from "./base.ts";
import { escapeTelnetData, TELNET, TelnetBuffer } from "./telnet.ts";

/** The socket this transport drives. */
export interface TelnetSocket {
  /** Whether the socket is closing or closed. */
  readonly closing: boolean;
  /** Up to `maxBytes`; an empty result means the far end has gone. */
  read(maxBytes: number): Promise<Uint8Array>;
  /** Put bytes on the wire. */
  write(data: Uint8Array): Promise<void>;
  /** Close it. */
  close(): Promise<void>;
  /** The address actually reached, for the post-connect egress check. */
  peerAddress(): string | undefined;
}

/** How a socket is opened. */
export type TelnetConnect = (host: string, port: number, timeoutMs: number) => Promise<TelnetSocket>;

/** Options for {@link TelnetTransport}. */
export interface TelnetTransportOptions {
  /** Opens the socket. Injected, so a test needs no network. */
  connect: TelnetConnect;
}

/** Per-connection options. */
export interface TelnetConnectOptions {
  /** Terminal width, offered through NAWS. */
  cols?: number;
  /** Terminal height, offered through NAWS. */
  rows?: number;
  /** Terminal type, offered through TTYPE. */
  term?: string;
  /** How long to wait for the connection. */
  timeoutS?: number;
}

/** How long a connection attempt is given. */
export const TELNET_CONNECT_TIMEOUT_S = 30.0;

/**
 * How much unconsumed input is tolerated.
 *
 * A subnegotiation has no length bound, so an upstream that sends `IAC SB`
 * and never `IAC SE` would grow this without limit. The cap is far above any
 * legitimate one; reaching it means the peer is broken or hostile.
 */
export const TELNET_MAX_RX_BUFFER = 256 * 1024;

/** Terminal geometry offered when the caller names none. */
export const TELNET_TRANSPORT_DEFAULTS = { cols: 80, rows: 25, term: "ANSI" } as const;

/** The subnegotiation byte that says "this is my terminal type". */
const TTYPE_IS = 0;

/** The subnegotiation byte that asks for it. */
const TTYPE_SEND = 1;

/** Backspace, which a BBS deletes with. */
const BACKSPACE = 0x08;

/** Delete, which a browser terminal sends for the backspace key. */
const DELETE = 0x7f;

/** Which of the four commands an option has already been sent. */
type NegotiationSide = "will" | "wont" | "do" | "dont";

/** A full RFC 854 telnet client behind the transport interface. */
export class TelnetTransport implements ConnectionTransport {
  readonly #connect: TelnetConnect;
  #socket: TelnetSocket | undefined;
  #buffer = new TelnetBuffer();
  #pending = 0;
  #cols: number = TELNET_TRANSPORT_DEFAULTS.cols;
  #rows: number = TELNET_TRANSPORT_DEFAULTS.rows;
  #term: string = TELNET_TRANSPORT_DEFAULTS.term;
  /**
   * What has already been said about each option.
   *
   * Shared between what arrives and what is sent, exactly as the reference
   * does it: an incoming `WILL X` records `will[X]`, so this end will not
   * then send its own `WILL X`. That is what stops two polite implementations
   * negotiating at each other forever.
   */
  readonly #negotiated: Record<NegotiationSide, Set<number>> = {
    will: new Set(),
    wont: new Set(),
    do: new Set(),
    dont: new Set(),
  };

  constructor(options: TelnetTransportOptions) {
    this.#connect = options.connect;
  }

  /**
   * Open the connection and make the opening offer.
   *
   * @throws {TransportConnectionError} If the socket cannot be opened.
   */
  async connect(host: string, port: number, options: TelnetConnectOptions = {}): Promise<void> {
    if (this.#socket !== undefined) {
      await this.disconnect();
    }
    const timeoutS = options.timeoutS ?? TELNET_CONNECT_TIMEOUT_S;
    try {
      this.#socket = await this.#connect(host, port, timeoutS * 1000);
    } catch (error) {
      throw new TransportConnectionError(`Failed to connect to ${host}:${port}`, { cause: error });
    }
    this.#buffer = new TelnetBuffer();
    this.#pending = 0;
    this.#cols = options.cols ?? TELNET_TRANSPORT_DEFAULTS.cols;
    this.#rows = options.rows ?? TELNET_TRANSPORT_DEFAULTS.rows;
    this.#term = options.term ?? TELNET_TRANSPORT_DEFAULTS.term;
    // Binary and suppress-go-ahead, unasked: without them the session is
    // line-mode and seven-bit, and a BBS draws for the wrong terminal.
    await this.#sendOnce("will", TELNET.OPT_BINARY);
    await this.#sendOnce("will", TELNET.OPT_SGA);
  }

  /** Close the connection. Safe to call twice. */
  async disconnect(): Promise<void> {
    const socket = this.#socket;
    if (socket === undefined) {
      return;
    }
    this.#socket = undefined;
    try {
      await socket.close();
    } catch {
      // Cleanup that raised would leave the transport wedged as connected.
    }
  }

  /**
   * Send bytes, escaped for the wire.
   *
   * @throws {TransportConnectionError} If not connected, or the far end went
   *   mid-send.
   */
  async send(data: Uint8Array): Promise<void> {
    const socket = this.#requireSocket();
    // A browser terminal sends DEL for the backspace key; a BBS deletes with
    // backspace, so a literal DEL prints as a stray character instead.
    const remapped = Uint8Array.from(data, (byte) => (byte === DELETE ? BACKSPACE : byte));
    try {
      await socket.write(escapeTelnetData(remapped));
    } catch (error) {
      await this.disconnect();
      throw new TransportConnectionError("Connection lost", { cause: error });
    }
  }

  /**
   * Read application bytes, answering any negotiation that arrives.
   *
   * @throws {TransportConnectionError} If not connected, the far end closed
   *   with nothing left to hand over, or the peer sent more unconsumed input
   *   than {@link TELNET_MAX_RX_BUFFER}.
   */
  async receive(maxBytes: number, timeoutMs: number): Promise<Uint8Array> {
    const socket = this.#requireSocket();
    let chunk: Uint8Array | typeof TIMED_OUT;
    try {
      chunk = await this.#readWithTimeout(socket, maxBytes, timeoutMs);
    } catch (error) {
      await this.disconnect();
      throw new TransportConnectionError("Connection lost", { cause: error });
    }
    if (chunk === TIMED_OUT) {
      // A quiet terminal is not a broken one.
      return new Uint8Array(0);
    }

    if (chunk.length === 0) {
      // The far end has gone. Anything held back as an incomplete sequence is
      // handed over rather than dropped, and only then is the connection torn
      // down.
      const flushed = this.#buffer.flush();
      await this.disconnect();
      if (flushed.payload.length > 0) {
        return flushed.payload;
      }
      throw new TransportConnectionError("Connection closed by remote");
    }

    const result = this.#buffer.feed(chunk);
    this.#pending = this.#buffer.pending;
    if (this.#pending > TELNET_MAX_RX_BUFFER) {
      this.#buffer = new TelnetBuffer();
      this.#pending = 0;
      throw new TransportConnectionError(
        `telnet receive buffer exceeded ${TELNET_MAX_RX_BUFFER} bytes ` +
          "(likely IAC SB without IAC SE) — closing connection",
      );
    }
    // Answered before returning rather than in a background task: the bytes
    // that reach the wire are the same either way, and doing it here means a
    // caller that reads once has already replied by the time it looks.
    for (const event of result.events) {
      if (event.kind === "negotiate") {
        await this.#answerNegotiation(event.command, event.option);
      } else {
        await this.#answerSubnegotiation(event.payload);
      }
    }
    return result.payload;
  }

  /** Whether the connection is live. */
  isConnected(): boolean {
    return this.#socket !== undefined && !this.#socket.closing;
  }

  /**
   * The address actually reached.
   *
   * The real peer, not the hostname asked for, so an egress check runs
   * against what the connection went to. Absent means "proceed": a caller
   * must not fail closed on a socket that simply cannot say.
   */
  peerIp(): string | undefined {
    return this.#socket?.peerAddress();
  }

  /**
   * Tell the far end the terminal changed size.
   *
   * @throws {TransportConnectionError} If not connected.
   */
  async setSize(cols: number, rows: number): Promise<void> {
    this.#requireSocket();
    this.#cols = cols;
    this.#rows = rows;
    await this.#sendNaws(cols, rows);
  }

  /** Read, or give up after `timeoutMs`. */
  async #readWithTimeout(
    socket: TelnetSocket,
    maxBytes: number,
    timeoutMs: number,
  ): Promise<Uint8Array | typeof TIMED_OUT> {
    let timer: ReturnType<typeof setTimeout> | undefined;
    const timeout = new Promise<typeof TIMED_OUT>((resolve) => {
      timer = setTimeout(() => resolve(TIMED_OUT), timeoutMs);
    });
    try {
      return await Promise.race([socket.read(maxBytes), timeout]);
    } finally {
      clearTimeout(timer);
    }
  }

  /** The live socket, or a refusal. */
  #requireSocket(): TelnetSocket {
    if (this.#socket === undefined) {
      throw new TransportConnectionError("Not connected");
    }
    return this.#socket;
  }

  /**
   * Answer one negotiation.
   *
   * The reference guards against the socket having gone, because it dispatches
   * the answer as a background task that may run after a disconnect. Answering
   * inline means that cannot happen here, so there is no guard to write — the
   * write itself still checks, for the same reason.
   */
  async #answerNegotiation(command: number, option: number): Promise<void> {
    // What arrives is recorded in the same table the sends check, so this end
    // does not answer an option the far end has already settled.
    this.#record(command, option);
    if (command === TELNET.DO) {
      await this.#answerDo(option);
      return;
    }
    if (command === TELNET.WILL) {
      // Echo, suppress-go-ahead and binary are the three worth having.
      await this.#sendOnce(
        option === TELNET.OPT_ECHO || option === TELNET.OPT_SGA || option === TELNET.OPT_BINARY ? "do" : "dont",
        option,
      );
      return;
    }
    // A refusal in either direction is answered with the matching refusal.
    await this.#sendOnce(command === TELNET.DONT ? "wont" : "dont", option);
  }

  /** Answer a DO, sending the value where the option needs one. */
  async #answerDo(option: number): Promise<void> {
    if (option === TELNET.OPT_BINARY || option === TELNET.OPT_SGA) {
      await this.#sendOnce("will", option);
      return;
    }
    if (option === TELNET.OPT_NAWS) {
      await this.#sendOnce("will", option);
      // Agreeing and then never saying how big leaves the server guessing.
      await this.#sendNaws(this.#cols, this.#rows);
      return;
    }
    if (option === TELNET.OPT_TTYPE) {
      await this.#sendOnce("will", option);
      await this.#sendTtype(this.#term);
      return;
    }
    await this.#sendOnce("wont", option);
  }

  /** Answer a subnegotiation, which in practice is a terminal-type request. */
  async #answerSubnegotiation(payload: Uint8Array): Promise<void> {
    if (payload.length > 1 && payload[0] === TELNET.OPT_TTYPE && payload[1] === TTYPE_SEND) {
      await this.#sendTtype(this.#term);
    }
  }

  /** Record what has been said about an option. */
  #record(command: number, option: number): void {
    const side =
      command === TELNET.DO ? "do" : command === TELNET.DONT ? "dont" : command === TELNET.WILL ? "will" : "wont";
    this.#negotiated[side].add(option);
  }

  /**
   * Send a command for an option, at most once.
   *
   * Two polite implementations that answered every message would negotiate at
   * each other forever.
   */
  async #sendOnce(side: NegotiationSide, option: number): Promise<void> {
    if (this.#negotiated[side].has(option)) {
      return;
    }
    const command =
      side === "will" ? TELNET.WILL : side === "wont" ? TELNET.WONT : side === "do" ? TELNET.DO : TELNET.DONT;
    await this.#writeRaw(Uint8Array.from([TELNET.IAC, command, option]));
    this.#negotiated[side].add(option);
  }

  /** Send the window size. */
  async #sendNaws(cols: number, rows: number): Promise<void> {
    const size = Uint8Array.from([(cols >> 8) & 0xff, cols & 0xff, (rows >> 8) & 0xff, rows & 0xff]);
    await this.#writeSubnegotiation(TELNET.OPT_NAWS, size);
  }

  /** Send the terminal type. */
  async #sendTtype(term: string): Promise<void> {
    const name = Uint8Array.from(term, (character) => character.charCodeAt(0) & 0xff);
    await this.#writeSubnegotiation(TELNET.OPT_TTYPE, Uint8Array.from([TTYPE_IS, ...name]));
  }

  /** Frame and send a subnegotiation. */
  async #writeSubnegotiation(option: number, payload: Uint8Array): Promise<void> {
    // Every 0xFF inside the payload is doubled, or the receiver reads it as
    // the framing byte and the block ends in the wrong place.
    await this.#writeRaw(
      Uint8Array.from([TELNET.IAC, TELNET.SB, option, ...escapeTelnetData(payload), TELNET.IAC, TELNET.SE]),
    );
  }

  /** Write without escaping — these are commands, not data. */
  async #writeRaw(data: Uint8Array): Promise<void> {
    const socket = this.#socket;
    if (socket === undefined || socket.closing) {
      return;
    }
    try {
      await socket.write(data);
    } catch {
      // A negotiation that cannot be sent is not worth failing the read that
      // triggered it; the next operation will find the socket gone.
    }
  }
}

/**
 * Distinguishes a timed-out read from an empty one.
 *
 * They mean opposite things — nothing has arrived yet, versus the far end has
 * gone — so they cannot both be an empty buffer.
 */
const TIMED_OUT = Symbol("timed out");

File: packages/provide-uterm-client/src/provide/uterm/transports/telnet_transport.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""TelnetTransport — full RFC 854 client implementing ConnectionTransport."""

from __future__ import annotations

import asyncio
import contextlib
from typing import TYPE_CHECKING, Any

from provide.telemetry import get_logger

from provide.uterm.transports._telnet_const import (
    DO,
    DONT,
    IAC,
    OPT_BINARY,
    OPT_ECHO,
    OPT_NAWS,
    OPT_SGA_OPT,
    OPT_TTYPE,
    SB,
    SE,
    TTYPE_IS,
    WILL,
    WONT,
)

# Re-exported for callers that import these names from this module; they are
# part of this module's public surface but are not referenced internally.
from provide.uterm.transports._telnet_const import (
    ECHO as ECHO,
)
from provide.uterm.transports._telnet_const import (
    NAWS as NAWS,
)
from provide.uterm.transports._telnet_const import (
    SGA as SGA,
)
from provide.uterm.transports.base import ConnectionTransport

if TYPE_CHECKING:
    from asyncio import StreamReader, StreamWriter

logger = get_logger(__name__)

# Default connection timeout in seconds
_DEFAULT_CONNECT_TIMEOUT_S: float = 30.0

# Max unconsumed receive-buffer bytes before the connection is treated as
# hostile/malformed. An upstream that sends `IAC SB` without `IAC SE` would
# otherwise grow _rx_buf without bound (memory-exhaustion DoS). 256 KiB is far
# above any legitimate telnet subnegotiation.
_MAX_RX_BUF_BYTES: int = 256 * 1024


class TelnetTransport(ConnectionTransport):
    """Full RFC 854 telnet client implementing the ConnectionTransport interface.

    Unlike :class:`TelnetClient`, this class:

    - Implements :class:`~provide.uterm.transports.base.ConnectionTransport`.
    - IAC-escapes ``0xFF`` bytes in outgoing data (binary safety).
    - Handles full option negotiation: ECHO, SGA, NAWS, TTYPE.
    - Buffers incoming data and strips IAC sequences before returning payload.

    Example::

        transport = TelnetTransport()
        await transport.connect("bbs.example.com", 23, cols=80, rows=25)
        await transport.send(b"hello\\r")
        data = await transport.receive(4096, timeout_ms=5000)
        await transport.disconnect()
    """

    def __init__(self) -> None:
        self._reader: StreamReader | None = None
        self._writer: StreamWriter | None = None
        self._negotiated: dict[str, set[int]] = {"do": set(), "dont": set(), "will": set(), "wont": set()}
        self._rx_buf = bytearray()
        self._cols: int = 80
        self._rows: int = 25
        self._term: str = "ANSI"
        self._tasks: set[asyncio.Task[None]] = set()

    @staticmethod
    def _find_subneg_end(buf: bytes, start: int) -> int | None:
        """Find the end of a SB...SE subnegotiation block.

        Returns the index just after SE, or None if the block is incomplete.
        """
        j = start
        while j < len(buf) - 1:
            if buf[j] == IAC and buf[j + 1] == SE:
                return j + 2
            j += 1
        return None

    @staticmethod
    def _parse_telnet_buffer(
        data: bytes | bytearray, final: bool = False
    ) -> tuple[bytes, list[tuple[str, int, int | bytes]], int]:
        """Parse complete telnet sequences from a buffer.

        Returns application payload bytes, control events, and bytes consumed.
        Trailing incomplete sequences are left unconsumed by the caller
        unless final=True.
        """
        result = bytearray()
        events: list[tuple[str, int, int | bytes]] = []
        i = 0
        consumed = 0
        buf = bytes(data)

        while i < len(buf):
            if buf[i] != IAC:
                result.append(buf[i])
                i += 1
                consumed = i
                continue

            if i + 1 >= len(buf):
                if final:
                    result.append(IAC)
                    i += 1
                    consumed = i
                break

            cmd = buf[i + 1]
            if cmd in (DO, DONT, WILL, WONT):
                if i + 2 >= len(buf):
                    if final:
                        # Truncated negotiation: emit as literal data
                        result.extend(buf[i:])
                        i = len(buf)
                        consumed = i
                    break
                events.append(("negotiate", cmd, buf[i + 2]))
                i += 3
                consumed = i
                continue

            if cmd == SB:
                end = TelnetTransport._find_subneg_end(buf, i + 2)
                if end is None:
                    if final:
                        # Truncated subnegotiation: emit as literal data
                        result.extend(buf[i:])
                        i = len(buf)
                        consumed = i
                    break
                payload = buf[i + 2 : end - 2]
                events.append(("subnegotiation", 0, payload))
                i = end
                consumed = i
                continue

            if cmd == IAC:
                result.append(IAC)
                i += 2
                consumed = i
                continue

            i += 2
            consumed = i

        return bytes(result), events, consumed

    def _consume_rx_buffer(self, final: bool = False) -> tuple[bytes, list[tuple[str, int, int | bytes]]]:
        payload, events, consumed = self._parse_telnet_buffer(self._rx_buf, final=final)
        if consumed:
            del self._rx_buf[:consumed]
        return payload, events

    async def connect(
        self,
        host: str,
        port: int,
        cols: int = 80,
        rows: int = 25,
        term: str = "ANSI",
        timeout: float = _DEFAULT_CONNECT_TIMEOUT_S,
        **_kwargs: Any,
    ) -> None:
        """Open a telnet connection.

        Args:
            host: Remote hostname or IP.
            port: Remote port.
            cols: Terminal columns for NAWS.
            rows: Terminal rows for NAWS.
            term: Terminal type string (e.g. ``"ANSI"``).
            timeout: Connection timeout in seconds.

        Raises:
            ConnectionError: If the connection attempt fails.
        """
        if self._writer:
            await self.disconnect()
        try:
            self._reader, self._writer = await asyncio.wait_for(asyncio.open_connection(host, port), timeout=timeout)
        except Exception as exc:
            raise ConnectionError(f"Failed to connect to {host}:{port}") from exc

        self._cols = cols
        self._rows = rows
        self._term = term

        await self._send_will(OPT_BINARY)
        await self._send_will(OPT_SGA_OPT)
        logger.debug("telnet_transport connected host=%s port=%d", host, port)

    async def disconnect(self) -> None:
        """Close the connection."""
        if not self._writer:
            return
        try:
            self._writer.close()
            await self._writer.wait_closed()
        except (ConnectionResetError, BrokenPipeError, RuntimeError):
            pass
        finally:
            for t in list(self._tasks):
                t.cancel()
            self._tasks.clear()
            self._writer = None
            self._reader = None
            self._rx_buf.clear()
            self._negotiated = {"do": set(), "dont": set(), "will": set(), "wont": set()}

    async def send(self, data: bytes) -> None:
        """Send bytes with IAC escaping (RFC 854: ``0xFF`` → ``0xFF 0xFF``).

        Args:
            data: Raw bytes to send.

        Raises:
            ConnectionError: If not connected.
        """
        if not self._writer:
            raise ConnectionError("Not connected")
        # Remap DEL (0x7F) → BS (0x08): xterm.js sends DEL for Backspace,
        # but many BBS/telnet servers expect BS for character deletion.
        escaped = data.replace(b"\x7f", b"\x08").replace(b"\xff", b"\xff\xff")
        try:
            self._writer.write(escaped)
            await self._writer.drain()
        except (ConnectionResetError, BrokenPipeError) as exc:
            await self.disconnect()
            raise ConnectionError("Send failed") from exc

    @staticmethod
    async def _read_bounded(reader: StreamReader, max_bytes: int, timeout: float) -> bytes:
        """Read up to *max_bytes*, raising :class:`TimeoutError` after *timeout*.

        Deliberately not ``asyncio.wait_for``. On CPython < 3.12 ``wait_for``
        can *swallow* a ``CancelledError`` when the inner future completes in
        the same event-loop tick as the cancellation (gh-86296): it returns the
        read result instead of propagating, so a caller polling in a loop keeps
        running after it was cancelled and never releases the connection.
        ``asyncio.wait`` propagates cancellation unconditionally; the ``finally``
        makes sure the in-flight read is dropped on both the timeout and the
        cancellation path.
        """
        read_task = asyncio.ensure_future(reader.read(max_bytes))
        try:
            done, _pending = await asyncio.wait({read_task}, timeout=timeout)
            if not done:
                raise TimeoutError
            return read_task.result()
        finally:
            if not read_task.done():
                read_task.cancel()

    async def receive(self, max_bytes: int, timeout_ms: int) -> bytes:
        """Receive bytes, stripping IAC sequences.

        Args:
            max_bytes: Max bytes to read.
            timeout_ms: Read timeout in milliseconds (0 means return immediately on no data).

        Returns:
            Application-layer bytes (may be empty on timeout).

        Raises:
            ConnectionError: If not connected or connection closed.
        """
        if not self._reader:
            raise ConnectionError("Not connected")
        try:
            chunk = await self._read_bounded(self._reader, max_bytes, timeout_ms / 1000)
        except TimeoutError:
            return b""
        except (ConnectionResetError, BrokenPipeError) as exc:
            await self.disconnect()
            raise ConnectionError("Connection lost") from exc

        if not chunk:
            payload, events = self._consume_rx_buffer(final=True)
            await self.disconnect()
            if payload:
                return payload
            raise ConnectionError("Connection closed by remote")

        self._rx_buf.extend(chunk)
        payload, events = self._consume_rx_buffer()
        if len(self._rx_buf) > _MAX_RX_BUF_BYTES:
            self._rx_buf.clear()
            raise ConnectionError(
                f"telnet receive buffer exceeded {_MAX_RX_BUF_BYTES} bytes "
                "(likely IAC SB without IAC SE) — closing connection"
            )
        for event_type, cmd, opt_or_payload in events:
            if event_type == "negotiate":
                task = asyncio.create_task(self._negotiate(cmd, int(opt_or_payload)))
            else:
                task = asyncio.create_task(self._handle_subnegotiation(bytes(opt_or_payload)))
            self._tasks.add(task)
            task.add_done_callback(self._tasks.discard)
        return payload

    def is_connected(self) -> bool:
        """Return ``True`` if the connection is active."""
        return self._writer is not None and not self._writer.is_closing()

    def peer_ip(self) -> str | None:
        """Return the connected peer's IP address, or None if unavailable.

        Reads ``get_extra_info("peername")`` from the underlying asyncio
        transport (the real IP we connected to, NOT the original hostname) for
        post-connect peer-IP egress validation (M3 DNS-rebinding mitigation).
        Returns None when there is no writer or no peername (closed/half-open) —
        callers treat None as "proceed", never fail-closed.
        """
        if self._writer is None:
            return None
        peer = self._writer.get_extra_info("peername")
        if not peer:
            return None
        return str(peer[0])

    async def set_size(self, cols: int, rows: int) -> None:
        """Update terminal size and send NAWS subnegotiation.

        Args:
            cols: New column count.
            rows: New row count.

        Raises:
            ConnectionError: If not connected.
        """
        if not self._writer:
            raise ConnectionError("Not connected")
        self._cols = cols
        self._rows = rows
        await self._send_naws(cols, rows)

    def _track_negotiation_state(self, cmd: int, opt: int) -> None:
        """Record the negotiation command in the appropriate tracking set."""
        if cmd == DO:
            self._negotiated["do"].add(opt)
        elif cmd == DONT:
            self._negotiated["dont"].add(opt)
        elif cmd == WILL:
            self._negotiated["will"].add(opt)
        elif cmd == WONT:  # pragma: no branch
            self._negotiated["wont"].add(opt)

    async def _negotiate_do_response(self, opt: int) -> None:
        """Send the appropriate response to a DO command."""
        if opt in (OPT_BINARY, OPT_SGA_OPT):
            await self._send_will(opt)
        elif opt == OPT_NAWS:
            await self._send_will(opt)
            await self._send_naws(self._cols, self._rows)
        elif opt == OPT_TTYPE:
            await self._send_will(opt)
            await self._send_ttype(self._term)
        else:
            await self._send_wont(opt)

    async def _negotiate_will_response(self, opt: int) -> None:
        """Send the appropriate response to a WILL command."""
        if opt in (OPT_ECHO, OPT_SGA_OPT, OPT_BINARY):
            await self._send_do(opt)
        else:
            await self._send_dont(opt)

    async def _negotiate(self, cmd: int, opt: int) -> None:
        if not self._writer:
            return
        self._track_negotiation_state(cmd, opt)
        try:
            if cmd == DO:
                await self._negotiate_do_response(opt)
            elif cmd == DONT:
                await self._send_wont(opt)
            elif cmd == WILL:
                await self._negotiate_will_response(opt)
            elif cmd == WONT:  # pragma: no branch
                await self._send_dont(opt)
        except (ConnectionResetError, BrokenPipeError):
            pass

    async def _handle_subnegotiation(self, sub: bytes) -> None:
        if not sub or not self._writer:
            return
        if sub[0] == OPT_TTYPE and len(sub) > 1 and sub[1] == 1:
            await self._send_ttype(self._term)

    async def _send_cmd(self, cmd: int, opt: int) -> None:
        if not self._writer or self._writer.is_closing():
            return
        self._writer.write(bytes([IAC, cmd, opt]))
        with contextlib.suppress(ConnectionResetError, BrokenPipeError):
            await self._writer.drain()

    async def _send_will(self, opt: int) -> None:
        # NOTE: _negotiate tasks run concurrently; two tasks could both pass the
        # `not in` check before either adds to the set (TOCTOU).  In practice
        # this only occurs if the server sends duplicate DO/WILL for the same
        # option, which is a protocol violation.  A duplicate WILL is harmless.
        if opt not in self._negotiated["will"]:
            await self._send_cmd(WILL, opt)
            self._negotiated["will"].add(opt)

    async def _send_wont(self, opt: int) -> None:
        if opt not in self._negotiated["wont"]:
            await self._send_cmd(WONT, opt)
            self._negotiated["wont"].add(opt)

    async def _send_do(self, opt: int) -> None:
        if opt not in self._negotiated["do"]:
            await self._send_cmd(DO, opt)
            self._negotiated["do"].add(opt)

    async def _send_dont(self, opt: int) -> None:
        if opt not in self._negotiated["dont"]:
            await self._send_cmd(DONT, opt)
            self._negotiated["dont"].add(opt)

    @staticmethod
    def _escape_iac(payload: bytes) -> bytes:
        """RFC 855/1073: every 0xFF byte inside a subnegotiation payload must
        be doubled so the receiver does not mis-parse it as an IAC framing byte.
        """
        return payload.replace(b"\xff", b"\xff\xff")

    async def _send_naws(self, cols: int, rows: int) -> None:
        if not self._writer or self._writer.is_closing():
            return
        wh = (cols >> 8) & 0xFF
        wl = cols & 0xFF
        hh = (rows >> 8) & 0xFF
        hl = rows & 0xFF
        # Escape 0xFF bytes in the window-size payload only (not the IAC SB/SE framing).
        size_bytes = self._escape_iac(bytes([wh, wl, hh, hl]))
        self._writer.write(bytes([IAC, SB, OPT_NAWS]) + size_bytes + bytes([IAC, SE]))
        with contextlib.suppress(ConnectionResetError, BrokenPipeError):
            await self._writer.drain()

    async def _send_ttype(self, term: str) -> None:
        payload = bytes([OPT_TTYPE, TTYPE_IS]) + term.encode("ascii", errors="replace")
        await self._send_subnegotiation(payload)

    async def _send_subnegotiation(self, payload: bytes) -> None:
        if not self._writer or self._writer.is_closing():
            return
        # Escape 0xFF bytes in the user payload only (not the IAC SB/SE framing).
        escaped_payload = self._escape_iac(payload)
        self._writer.write(bytes([IAC, SB]) + escaped_payload + bytes([IAC, SE]))
        with contextlib.suppress(ConnectionResetError, BrokenPipeError):
            await self._writer.drain()
